### PR TITLE
Remove unnecessary calls to Str::as_slice

### LINKED ISSUE #14433: core::str::Str.as_slice() should be renamed as_str()
The method `as_slice()` in the `core::str::Str` trait should be renamed to `as_str()`. This avoids the name collision with `core::slice::Vector.as_slice()` for types that need to implement both `Vector` and `Str` (as I believe `Vec<Ascii>` should).

I'm filing this issue now to give a chance for anyone to suggest any reasons why this is a bad idea. Otherwise I will go ahead and make the change myself.

### PR DESCRIPTION
**Pull request by richo**
This may not be quite ready to go out, I fixed some docs but suspect I missed a bunch.

I also wound up fixing a bunch of redundant `[]` suffixes, but on closer inspection I don't believe that can land until after a snapshot.

### CODE REVIEW

**Comment by richo** [comment]
Ok, I think I got everything, but no doubt I missed something in a test, on a platform I don't have to test on. A bors run would be helpful.

there were cases where Deref coercions would have worked, but I opted for `as_str` anyway, as it made the surrounding code more readable (eg, the fact that it was going to coerce to a str was not obvious)

**Comment by richo** [comment]
Following up from IRC, the places where I left `.as_str()`, eg the assertions were all places where it seemed that as a reader, explicitly taking the underlying str wasn't obvious.

More than happy to go back and fix them all

File: src/libcollections/fmt.rs [hunk 1]
```diff
@@ -443,6 +443,6 @@ mod tests {
     #[test]
     fn test_format() {
         let s = fmt::format(format_args!("Hello, {}!", "world"));
-        assert_eq!(s.as_slice(), "Hello, world!");
+        assert_eq!(s.as_str(), "Hello, world!");
```
Comment (Gankro) [inline src/libcollections/fmt.rs]: Places like these I think should be fine to drop the as_str from, clarity-wise.

File: src/librand/lib.rs [hunk 2]
```diff
@@ -298,9 +298,9 @@ pub trait Rng : Sized {
     /// let mut rng = thread_rng();
     /// let mut y = [1, 2, 3];
     /// rng.shuffle(&mut y);
-    /// println!("{:?}", y.as_slice());
+    /// println!("{:?}", y.as_str());
```
Comment (Gankro) [inline src/librand/lib.rs]: These as_str's seem necessary?

Also confusing, actually. [T; n] implements as_str?

File: src/librustc_bitflags/lib.rs [hunk 3]
```diff
@@ -83,7 +83,7 @@
 ///     let mut flags = FLAG_A | FLAG_B;
 ///     flags.clear();
 ///     assert!(flags.is_empty());
-///     assert_eq!(format!("{:?}", flags).as_slice(), "hi!");
+///     assert_eq!(format!("{:?}", flags).as_str(), "hi!");
```
Comment (Gankro) [inline src/librustc_bitflags/lib.rs]: I'd drop this as_str

File: src/librustc_borrowck/borrowck/mod.rs [hunk 4]
```diff
@@ -693,7 +693,7 @@ impl<'a, 'tcx> BorrowckCtxt<'a, 'tcx> {
             .span_err(span,
                       (format!("partial reinitialization of uninitialized \
                                structure `{}`",
-                               self.loan_path_to_string(lp))).as_slice());
+                               self.loan_path_to_string(lp))).as_str());
```
Comment (Gankro) [inline src/librustc_borrowck/borrowck/mod.rs]: What makes this format different from all the others that were changed?

File: src/libstd/env.rs [hunk 5]
```diff
@@ -860,7 +860,7 @@ mod tests {
         let n = make_rand_name();
         let s = repeat("x").take(10000).collect::<String>();
         set_var(&n, &s);
-        eq(var_os(&n), Some(s.as_slice()));
+        eq(var_os(&n), Some(s.as_str()));
```
Comment (Gankro) [inline src/libstd/env.rs]: Drop-worthy.

**Comment by Gankro** [comment]
Yeah going over this the as_str's don't really seem to add anything outside of a generic context. I would just drop 'em where you can.

**Comment by richo** [comment]
Thanks for all the feedback! I'll revise when I get a chance.

**Comment by alexcrichton** [comment]
Could you also leave `as_slice` in-place but `#[deprecated]` for now? I suspect there is a lot of code to go clean up outside the repo still using `.as_slice()` :)

File: src/librustc_borrowck/borrowck/mod.rs (hunk 4 above)
Comment (richo) [inline src/librustc_borrowck/borrowck/mod.rs]: I missed this one, although further poking seems to suggest that the &'s are unnecesary to get the String returned by format! to coerce

**Comment by nikomatsakis** [comment]
(Sadly, needs a rebase.)

**Comment by richo** [comment]
I redid this patch from scratch (I had some fairly lurky attempts to fix tests from the last time). Running make check now

**Comment by richo** [comment]
r? @Gankro I believe this is good to go out. the lldb tests are failing locally for me, but I believe it's because I have a wedged lldb install

(Error in case it's not something I've done is `Fatal Python error: PyThreadState_Get: no current thread` which then aborts python)

**Comment by richo** [comment]
Aaaaaand all tests pass on linux \o/

File: src/libcollections/str.rs [hunk 6]
```diff
@@ -96,11 +96,11 @@ impl<S: Str> SliceConcatExt<str, String> for [S] {
         }
 
         // `len` calculation may overflow but push_str will check boundaries
-        let len = s.iter().map(|s| s.as_slice().len()).sum();
+        let len = s.iter().map(|s| s.as_str().len()).sum();
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: This could be replaced with `&s[..]`

File: src/libcollections/str.rs [hunk 7]
```diff
@@ -96,11 +96,11 @@ impl<S: Str> SliceConcatExt<str, String> for [S] {
...
         let mut result = String::with_capacity(len);
 
         for s in s {
-            result.push_str(s.as_slice())
+            result.push_str(s.as_str())
```
Comment (alexcrichton) [inline src/libcollections/str.rs]: This could just be `s` (omitting the `as_str` I believe)

Comment (alexcrichton) [inline src/libcollections/str.rs]: Er nevermind, forgot that this was behind a `S: Str` bound (and above).

**Comment by alexcrichton** [comment]
It's still a little unclear to me whether we want to do this or not. I think that if generic conversion traits land then there's no real need for the `Str` trait or the `as_str` method as they're basically just specializations of that trait.

I think it'd be totally fine, however, to land as much removal of `as_str` as possible for now though!

File: src/libcollections/str.rs (hunk 6 above)
Comment (richo) [inline src/libcollections/str.rs]: There are a bunch of warnings about places I could use slice syntax, I figured I'd do this in two passes to make landing it feasible, although this was less churn than I expected.

I also wasn't sure if the syntax was dependant on a snapshot.

**Comment by richo** [comment]
Ok, I'll probably have a PR for that subset of this one in the next few days.

File: src/libcollections/str.rs (hunk 6 above)
Comment (alexcrichton) [inline src/libcollections/str.rs]: Yeah the slicing syntax should definitely be usable as-is today, thankfully it's been around for awhile!

**Comment by alexcrichton** [comment]
Thanks @richo!

**Comment by richo** [comment]
Thanks to totally unforeseen organisation on my part, the rebase to only remove `as_slice()`s was totally trivial. r? @alexcrichton

**Comment by alexcrichton** [comment]
Nice! I updated the title and removed the `Closes #xxxx` in the title, otherwise looks great to me!

**Comment by richo** [comment]
Great catch, cheers!

**Comment by Manishearth** [comment]
```
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/libstd/process.rs:833:32: 835:53 error: the trait `core::ops::Fn<(char,)>` is not implemented for the type `collections::string::String` [E0277]
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/libstd/process.rs:833                 assert!(output.contains(format!("{}={}",
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/libstd/process.rs:834                                                 *k,
/home/rustbuild/src/rust-buildbot/slave/auto-linux-64-x-android-t/build/src/libstd/process.rs:835  
```

**Comment by richo** [comment]
Test fixed, r?

**Comment by richo** [comment]
@Manishearth you might want to remove this from the rollup, I was just eyeballing it and spotted a few things that look off, I think it got mangled in the rebase. I have make check running here, but I don't want to toast another rollup :(

**Comment by Manishearth** [comment]
(already was removed :smile: )

**Comment by richo** [comment]
Ok, make check passes. I'm planning to hand hold this through bors, since I don't have access to an android test environment. (Fingers crossed I caught everything this time though)

**Pull request merged**